File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/3085_사탕게임/3085.py
N = int(input())
board = [list(input()) + [0] for _ in range(N)]
board.append([0] * (N + 1))
res = 1

def calc_col_score(col):
    global res

    if col >= N:
        return
    
    max_score = 1
    tmp = 1
    color = board[0][col]
    for i in range(1, N):
        if board[i][col] != color:
            max_score = max(max_score, tmp)
            tmp = 1
            color = board[i][col]
        else:
            tmp += 1
    res = max(res, max(max_score, tmp))

def calc_row_score(row):
    global res

    if row >= N:
        return

    max_score = 1
    tmp = 1
    color = board[row][0]
    for i in range(1, N):
        if board[row][i] != color:
            max_score = max(max_score, tmp)
            tmp = 1
            color = board[row][i]
        else:
            tmp += 1
    res = max(res, max(max_score, tmp))

def swap_candy(row, col):
    color = board[row][col]

    if color != board[row + 1][col]:
        board[row][col], board[row + 1][col] = board[row + 1][col], board[row][col]
        calc_row_score(row)
        calc_row_score(row + 1)
        calc_col_score(col)
        board[row][col], board[row + 1][col] = board[row + 1][col], board[row][col]
    if color != board[row][col + 1]:
        board[row][col], board[row][col + 1] = board[row][col + 1], board[row][col]
        calc_col_score(col)
        calc_col_score(col + 1)
        calc_row_score(row)
        board[row][col], board[row][col + 1] = board[row][col + 1], board[row][col]

for i in range(N):
    calc_row_score(i)
    calc_col_score(i)

for i in range(N):
    for j in range(N):
        swap_candy(i, j)

print(res)

File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/3085_사탕게임/3085.cpp
#include <algorithm>
#include <stdio.h>
#include <assert.h>

using namespace std;

int N, res = -1;
char board[55][55];

void calc_col_score(int col)
{
    if (col >= N)
        return;
        
    char color;
    int max_score, tmp;

    max_score = 1;
    tmp = 1;
    color = board[0][col];
    for (int i = 1; i < N; i++)
    {
        if (board[i][col] != color)
        {
            max_score = max(max_score, tmp);
            tmp = 1;
            color = board[i][col];
        }
        else
            ++tmp;
    }
    res = max(res, max(max_score, tmp));
}

void calc_row_score(int row)
{
    if (row >= N)
        return;

    char color;
    int max_score, tmp;

    max_score = 1;
    tmp = 1;
    color = board[row][0];
    for (int i = 1; i < N; i++)
    {
        if (board[row][i] != color)
        {
            max_score = max(max_score, tmp);
            tmp = 1;
            color = board[row][i];
        }
        else
            ++tmp;
    }
    res = max(res, max(max_score, tmp));
}

void swap_candy(int row, int col)
{
    char color = board[row][col];

    if (color != board[row + 1][col])
    {
        swap(board[row][col], board[row + 1][col]);
        calc_row_score(row);
        calc_row_score(row + 1);
        calc_col_score(col);
        swap(board[row][col], board[row + 1][col]);
    }
    if (color != board[row][col + 1])
    {
        swap(board[row][col], board[row][col + 1]);
        calc_col_score(col);
        calc_col_score(col + 1);
        calc_row_score(row);
        swap(board[row][col], board[row][col + 1]);
    }
}

int main()
{
    scanf("%d", &N);
    for (int i = 0; i < N; i++)
        scanf("%s", board[i]);
    
    for (int i = 0; i < N; i++)
    {
        calc_row_score(i);
        calc_col_score(i);
    }

    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            swap_candy(i, j);

    printf("%d\n", res);
    return 0;
}
